`kani-cov`: A coverage tool for Kani

adpaco-aws opened the pull request:
This PR introduces `kani-cov`, a coverage-oriented tool for Kani. This new tool will help users aggregate raw coverage results produced by Kani and generate human-readable coverage summaries and reports.

`kani-cov` allows using three different subcommands: `merge`, `summary` and `report`. In the following, we will show how each one of these commands can be used.

### The `merge` subcommand

The `merge` subcommand takes raw coverage results from Kani (AKA "kaniraw" files) and generates aggregated coverage results which are stored into another file (AKA a "kanicov" file). This new file can then be used for generating summaries or reports.

For example, if we assume this Rust code is in `main.rs`:

```rust
1 fn _other_function() {
2    println!("Hello, world!");
3 }
4
5 fn test_cov(val: u32) -> bool {
6     if val < 3 || val == 42 {
7         true
8     } else {
9         false
10    }
11 }
12
13 #[cfg_attr(kani, kani::proof)]
14 fn main() {
15    let test1 = test_cov(1);
16    let test2 = test_cov(2);
17    assert!(test1);
18    assert!(test2);
19 }
```
and we execute Kani with coverage enabled
```
kani main.rs --coverage -Zsource-coverage
```
the raw coverage results will be saved to a folder
```
[info] Coverage results saved to /home/ubuntu/coverage-experiments/src/kanicov_2024-09-23_23-49
```
we now can aggregate those results with the `merge` subcommand
```
kani-cov merge kanicov_2024-09-23_23-49/*kaniraw.json
```
which by default produces a `default_kanicov.json` file. We can also use the `--output <path>` option to change the name of the file.

The type of file we just produced is known as the "profile" or the "kanicov" file, and contains aggregated coverage results which may have been originated in multiple verification sessions. This file is a required input for the other two subcommands.

### The `summary` subcommand

Now that we have both the "kanicov" file and the "kanimap" file, we are ready to produce coverage metrics with the `summary` subcommand:

```
kani-cov summary kanicov_2024-09-23_23-49/kanicov_2024-09-23_23-49_kanimap.json --profile default_kanicov.json
```

This outputs a table including coverage metrics for functions, lines and regions for each file referenced in the coverage mappings:

```
| Filename                                      | Function (%) | Line (%)     | Region (%)  |
| --------------------------------------------- | ------------ | ------------ | ----------- |
| /home/ubuntu/coverage-experiments/src/main.rs | 2/3 (66.67)  | 8/12 (66.67) | 4/6 (66.67) |
```

Since the default format is markdown, the table can also be rendered as in this PR description:

| Filename                                      | Function (%) | Line (%)     | Region (%)  |
| --------------------------------------------- | ------------ | ------------ | ----------- |
| /home/ubuntu/coverage-experiments/src/main.rs | 2/3 (66.67)  | 8/12 (66.67) | 4/6 (66.67) |

At present, no other formats are available. But the JSON and CSV formats would be really nice and easy to have.

### The `report` subcommand

Coverage reports go a little further and allow users to visualize coverage information on the source code.

The `report` requires the same input files as the `summary` subcommand:

```
kani-cov report kanicov_2024-09-23_23-49/kanicov_2024-09-23_23-49_kanimap.json --profile default_kanicov.json
```

When calling this command from a terminal, it will render like this:

![Screenshot 2024-10-04 at 5 49 51 PM](https://github.com/user-attachments/assets/65c12d62-4dd8-40e6-bfe9-8fc0e70127d2)

However, if you're redirecting the output to a file, it will back off to another text-based mode which simply uses ` ``` ` and ` ''' ` to represent opening and closing escapes for the coverage regions. This is just another format called `escapes` which can be explicitly requested as

```
kani-cov report kanicov_2024-09-23_23-49/kanicov_2024-09-23_23-49_kanimap.json --profile default_kanicov.json --format=escapes
```

Similar to the `summary` command, this will iterate over the files in the coverage mappings, compute coverage information and highlight any uncovered regions (except #3543).

### Testing

This PR also changes `compiletest` to automatically call the `merge` and `report` subcommands on the `coverage` test suite. The tests have been blessed using the `--fix-expected` option and this Python3 script that removes the first line (path to the file) and adds `\` after each line so an exact match is expected.

```
import os

def process_files(directory):
    for filename in os.listdir(directory):
        if filename == "expected":
            file_path = os.path.join(directory, filename)
            with open(file_path, "r") as file:
                lines = file.readlines()
            
            # Remove the first line
            lines = lines[1:]
            
            # Add the character `\` after every line
            processed_lines = []
            for i, line in enumerate(lines):
                processed_lines.append(line.replace("\n", "\\\n"))
            
            # Write the processed lines back to the file
            with open(file_path, "w") as file:
                file.writelines(processed_lines)
```

### Callouts / Known issues

This PR only introduces `kani-cov`. It doesn't plug `kani-cov` into `kani-driver` so it's automatically run when using the `--coverage` option, nor it prepares the `kani-cov` binary for distribution. This is better to address later.

This PR also spawns a few issues:
 - #3543 forces us to filter out certain coverage results to avoid issues with the report command. It is important that we investigate where these coverage results come from and their semantics as this represents a soundness issue.
 - #3541 is to avoid duplicating certain data structures that currently live in `kani-driver`. We should probably have a different crate for these structures and import them from `kani-driver` and `kani-cov`, but this is more of a distribution problem that doesn't need to be tackled here.
 - #3542 requires us to match filenames when extracting function coverage results. It's not a big deal, but it'll likely improve the tool's robustness in the long term.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval reviewed (commented):
Are we planning to release this tool or is it only for developers?

celinval commented on tools/kanicov/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "visualize_coverage"
```
why not `kani-cov`?

adpaco-aws commented:
> Are we planning to release this tool or is it only for developers?

The plan is to release the tool. More details in the upcoming RFC.

adpaco-aws commented on Cargo.toml:
```diff
@@ -41,6 +41,7 @@ members = [
   "tools/bookrunner",
   "tools/compiletest",
   "tools/build-kani",
+  "tools/kanicov",
```
Thanks for the suggestion! Yes, let's use `kani-cov` instead.

adpaco-aws commented:
Note to self: Consider highlighting uncovered spans in red as the default behavior. At the very least, discuss in the RFC.

adpaco-aws commented:
Just marked this as ready for review!

feliperodri commented on tools/kani-cov/src/args.rs:
```diff
@@ -0,0 +1,111 @@
+/// arguments.
+pub fn validate_args(args: &Args) -> Result<()> {
+    if args.command.is_none() {
+        bail!("subcommand needs to be specified")
```
Nitpick: I'd include here the command options available.

adpaco-aws replied:
Done.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+    pub total_times: u32,
+}
+
+/// A coverage region.
```
I'd include
```suggestion
/// A coverage region.
/// `start` and `end` are tuples containing the line and column numbers.
```

adpaco-aws replied:
Done.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CoverageResults {
+    pub data: BTreeMap<String, Vec<CoverageCheck>>,
+}
+
+/// Aggregated coverage results.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CombinedCoverageResults {
+    pub data: BTreeMap<String, Vec<(String, Vec<CovResult>)>>,
+}
```
Wouldn't be better to use a `HashMap` here? Does the order matter? I also imagine this data could be quite large.

adpaco-aws replied:
The order doesn't really matter. Changed to `HashMap`.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+
+impl Display for CoverageRegion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{} - {}:{}", self.start.0, self.start.1, self.end.0, self.end.1)
```
Maybe let's include the file name as well.
```suggestion
        write!(f, "{}:{}:{} - {}:{}", self.file, self.start.0, self.start.1, self.end.0, self.end.1)
```

adpaco-aws replied:
Ok, done.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    let mut function_info: Vec<FunctionInfo> = Vec::new();
+
```
We could also declare here:
```rust
let source_bytes = source_code.as_bytes();
```
and avoid calling it multiple times.

adpaco-aws replied:
Yep, done.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+
+    let mut function_info: Vec<FunctionInfo> = Vec::new();
+
+    if cursor.node().kind() == "function_item" {
+        function_info.push(function_info_from_node(cursor.node(), source_code.as_bytes()))
+    };
+
+    while cursor.goto_next_sibling() {
+        if cursor.node().kind() == "function_item" {
+            function_info.push(function_info_from_node(cursor.node(), source_code.as_bytes()))
+        }
+    }
```
Why not use an iterator?
```suggestion
for node in tree.root_node().children(&mut cursor) {
    if node.kind() == "function_item" {
        function_info.push(function_info_from_node(node, source_bytes));
    }
}
```

adpaco-aws replied:
Thanks! Yes, this looks better.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+    let tree = parser.parse(&source_code, None).expect("Failed to parse file");
+
+    let mut cursor = tree.walk();
+    let first_child_exists = cursor.goto_first_child();
+
+    if !first_child_exists {
+        return vec![];
+    }
```
```suggestion
    if !cursor.goto_first_child() {
        return vec![];
    }
```

adpaco-aws replied:
This code is now gone because we're using the iterator.

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+    // The filenames in "kaniraw" files are not absolute, so we need to match
+    // them with the ones we have in the aggregated results (i.e., the filenames
+    // in the "kanimap" files).
+    let filename = file.to_path_buf().into_os_string().into_string().unwrap();
```
Wouldn't this also works? It's also safer since it returns valid UTF-8. 
```suggestion
    let filename = file.to_str()?;
```

adpaco-aws replied:
Yes, it works. I've updated the code to use your suggestion, thanks!

feliperodri commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+    // TODO: The filenames in kaniraw files should be absolute, just like in metadata.
+    // Otherwise the key for `results` just fails... <https://github.com/model-checking/kani/issues/3542>
+    let file_results = results.data.get(right_filename).unwrap();
+    let function = info.name.clone();
```
Why do you need a clone here? The reference might be fine.

adpaco-aws replied:
True, done.

feliperodri commented on tools/kani-cov/src/merge.rs:
```diff
@@ -0,0 +1,150 @@
+
+    let mut new_data: BTreeMap<String, Vec<(String, Vec<CovResult>)>> = BTreeMap::new();
+
+    for (file_name, fun_name) in all_file_function_names {
```
Wouldn't references be better here?
```suggestion
    for (file_name, fun_name) in &all_file_function_names {
```

adpaco-aws replied:
We can use references but that forces us to use a bunch of `*` and `clone()` in the code below. I didn't think it looked better than the current solution, but happy to commit it if you think otherwise.

feliperodri commented on tools/kani-cov/src/merge.rs:
```diff
@@ -0,0 +1,150 @@
+
+/// All function names appearing in raw coverage results
+fn function_names_from_results(results: &[CoverageResults]) -> Vec<(String, String)> {
+    let mut file_function_pairs = vec![];
+    for result in results {
+        let files = result.data.keys().cloned();
+        for file in files {
+            let checks = result.data.get(&file).unwrap();
+            for check in checks {
+                let function = check.function.clone();
+                let file_function = (file.clone(), function);
+                if !file_function_pairs.contains(&file_function) {
+                    file_function_pairs.push(file_function);
+                }
+            }
+        }
+    }
+    file_function_pairs
```
```suggestion
    let mut file_function_pairs = HashSet::new();
    for result in results {
        for (file, checks) in &result.data {
            for check in checks {
                file_function_pairs.insert((file.clone(), check.function.clone()));
            }
        }
    }
    file_function_pairs.into_iter().collect()
```

adpaco-aws replied:
Updated, thanks!

feliperodri commented on tools/kani-cov/src/report.rs:
```diff
@@ -0,0 +1,262 @@
+    filepath: PathBuf,
+    results: Vec<LineResults>,
+) -> Result<()> {
+    let flattened_results: Vec<(usize, Option<(u32, MarkerInfo)>)> =
```
Couldn't this be empty?

adpaco-aws replied:
Yes, this could be empty. For that to happen, there must be no functions in the file. But if there's at least one, and even if it's unreachable, `flattened_results` will have been populated with full-line markers.

feliperodri reviewed (changes requested):
Thank you so much for pushing this to the finish line. I can't wait to use it, only a few comments.

adpaco-aws commented on tools/compiletest/src/runtest.rs:
```diff
@@ -441,10 +441,68 @@ impl<'test> TestCx<'test> {
+    fn run_kanicov_report(
+        &self,
+        kanimap: &PathBuf,
+        kaniraws: &Vec<PathBuf>,
```
Done.

adpaco-aws commented on tools/compiletest/src/runtest.rs:
```diff
@@ -441,10 +441,68 @@ impl<'test> TestCx<'test> {
+        if coverage_info.is_none() {
+            self.fatal_proc_rec("failed to find the path to the coverage results", proc_res);
+        }
+        let coverage_path = coverage_info.unwrap().split(' ').last().unwrap();
```
Done.

adpaco-aws commented on tools/compiletest/src/runtest.rs:
```diff
@@ -441,10 +441,68 @@ impl<'test> TestCx<'test> {
+        (kanimap, kaniraws, kanicov)
     }
 
+    fn extract_cov_results_path(&self, proc_res: &ProcRes) -> PathBuf {
```
Because we need access to `fatal_proc_rec`, which is used for graceful failures in `compiletest`.

adpaco-aws commented on tools/kani-cov/src/args.rs:
```diff
@@ -0,0 +1,111 @@
+
+use crate::{merge, report, summary};
+
+/// We define three subcommands:
+///  * `merge` for merging raw Kani coverage results (AKA "kaniraw" files)
+///  * `summary` for producing a summary containing coverage metrics
+///  * `report` for generating human-readable coverage reports
```
Since the PR description already includes an example, I've done some modifications and added it here.

adpaco-aws commented on tools/kani-cov/src/args.rs:
```diff
@@ -0,0 +1,111 @@
+pub struct MergeArgs {
+    #[arg(long)]
+    pub output: Option<PathBuf>,
+    #[arg(required = true)]
+    pub files: Vec<PathBuf>,
```
I don't know, this seems to be working as intended. Why do you think we should add it to all required arguments?

adpaco-aws commented on tools/kani-cov/src/args.rs:
```diff
@@ -0,0 +1,111 @@
+pub enum SummaryFormat {
+    Markdown,
+    // Json,
+}
```
Done!

adpaco-aws commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,201 @@
+pub struct CovResult {
+    pub function: String,
+    pub region: CoverageRegion,
+    pub times_covered: u32,
```
True, and it's probably better to use `usize` anyway. I've replaced `u32` with `usize` in most types/structures, and added more documentation to `CovResult` regarding `total_times` and `times_covered`.

adpaco-aws commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module defines coverage-oriented data structures shared among
+//! subcommands and other utilities like the Rust tree-sitter.
```
This is a great suggestion, and it's now implemented.

feliperodri reviewed (changes requested):
Thanks for implementing all my comments. I have one last before approving it. Why the issue https://github.com/model-checking/kani/issues/3543 is classified as a soundness issue now? As far as I understood, the coverage report feature will crash in the cases you mentioned in the issue. Will it report false coverage reports in any case? This is important to fix before stabilization, but it seems to me unnecessary to classify this as a soundness issue.

carolynzech commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+type Function = String;
+type Filename = String;
+type LineNumber = usize;
+
```
```suggestion
type ColumnNumber = usize;

```

carolynzech commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+/// had the `COVERED` status.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CovResult {
+    pub function: String,
```
```suggestion
    pub function: Function,
```

carolynzech commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct CoverageRegion {
+    pub file: String,
```
```suggestion
    pub file: Filename,
```

carolynzech commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct CoverageRegion {
+    pub file: String,
+    pub start: (usize, usize),
+    pub end: (usize, usize),
```
```suggestion
    pub start: (LineNumber, ColumnNumber),
    pub end: (LineNumber, ColumnNumber),
```

carolynzech commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+#[derive(Debug)]
+pub struct FunctionInfo {
+    pub name: Function,
+    pub start: (usize, usize),
+    pub end: (usize, usize),
```
```suggestion
    pub start: (LineNumber, ColumnNumber),
    pub end: (LineNumber, ColumnNumber),
```

carolynzech commented on tools/kani-cov/src/coverage.rs:
```diff
@@ -0,0 +1,209 @@
+    info: &FunctionInfo,
+    file: &Path,
+    results: &CombinedCoverageResults,
+) -> Option<(String, Vec<CovResult>)> {
```
```suggestion
) -> Option<(Function, Vec<CovResult>)> {
```

carolynzech commented on tools/kani-cov/src/summary.rs:
```diff
@@ -0,0 +1,301 @@
+
+/// Computes coverage results from a line-based perspective.
+///
+/// Basically, for each line we produce an `<Option<(u32, MarkerInfo)>>` result
```
```suggestion
/// Basically, for each line we produce an `<Option<(usize, MarkerInfo)>>` result
```

carolynzech reviewed (approved):
I'd recommend making the type aliases public and incorporating them throughout `merge.rs` and `summary.rs` as well, but not a blocker for this PR.
Excited to use this feature!

adpaco-aws commented:
> Why the issue #3543 is classified as a soundness issue now? As far as I understood, the coverage report feature will crash in the cases you mentioned in the issue. Will it report false coverage reports in any case? This is important to fix before stabilization, but it seems to me unnecessary to classify this as a soundness issue.

The reason #3543 was classified as a soundness issue is because we were filtering out the results which could lead to a crash in `kani-cov`. Filtering out these results can be considered a soundness issue because, even if they're spurious failures coming from the Rust coverage instrumentation, we could be hiding other issues. It all depends on how we define soundness for coverage and the origin of these coverage artifacts.

Regardless, I've changed how we handle the checks with these out-of-bounds regions to replicate the spurious failure instead. Since we only need to extend the line by one space to represent the check in such cases, we just do that now. In other words, we are now reporting the same spurious failures, which may be more confusing for users but avoids soundness concerns. For example, for the test in `tests/coverage/assert` we get this report:

````
/home/ubuntu/kani/tests/coverage/assert/test.rs
   1|     | // Copyright Kani Contributors
   2|     | // SPDX-License-Identifier: Apache-2.0 OR MIT
   3|     | 
   4|     | #[kani::proof]
   5|    1| fn foo() {
   6|    1|     let x: i32 = kani::any();
   7|    1|     if x > 5 {
   8|     |         // fails
   9|    1|         assert!(x < 4);
  10|    1|         if x < 3 ```{'''
  11|    0| ```            // unreachable'''
  12|    0| ```            assert!(x == 2);'''
  13|    0| ```        }'''``` '''
  14|    1|     } else {
  15|    1|         // passes
  16|    1|         assert!(x <= 5);
  17|    1|     }
  18|     | }
````

Note the single region represented at the end of line 13? In the original source code, that space doesn't exist. But we can create that space in this representation and report the same result that the instrumentation gives us. It's less noticeable in the terminal output, but we also "draw it" there:

![Screenshot 2024-10-04 at 5 49 51 PM](https://github.com/user-attachments/assets/2190243b-b39f-4cc6-8b04-57e163f8810b)

In addition to this change, I've added some more documentation to the code handling this special case and updated #3543 (labeled as "spurious failure" now).  Please don't hesitate to ask for more clarifications if you need them. @carolynzech I've also implemented your suggestions to use the type aliases in other modules of `kani-cov`. Thanks!
